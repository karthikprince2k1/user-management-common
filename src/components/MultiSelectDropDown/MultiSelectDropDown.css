.dropdown {
  position: relative;
  display: inline-block;
  width: 300px;
}

.dropdown-list-hidden {
  display: none;
}

.dropdown-menu {
  padding: 0;
}

.dropdown-menu {
  width: 100%;
  padding-top: 5px;
  margin: 0;
  border: 1px solid #cccccc;
  height: 30px;
  text-align: left;
  overflow: hidden;
  border-radius: 4px;
  box-sizing: border-box;
}

.dropdown-selected-item {
  margin-left: 0px;
  margin-right: 3px;
  background-color: #0096fb;
  color: white;
  border-radius: 15px;
  text-align: center;
  padding: 3px 5px;
}
.dropdown-selected-item:hover {
  cursor: pointer;
}

.close::after {
  content: "\00d7";
}

.dropdown-menu svg {
  position: absolute;
  top: 2px;
  right: 3px;
}

.dropdown-menu svg.rotate {
  transform: rotate(180deg);
  padding: 0;
  margin: 0;
}

ul.dropdown-list-visible {
  display: block;
  text-align: left;
  padding: 0;
  margin: 0;
  border: 1px solid #cccccc;
  border-top: 0px;
  position: absolute;
  left: 0;
  width: 300px;
  box-sizing: border-box;
  list-style-type: none;
  border-radius: 4px;
  background: white;
  z-index: 1;
}

ul.dropdown-list-visible li {
  padding: 5px;
}
.dropdown-list-visible li:hover {
  cursor: pointer;
  background-color: #0096fb;
  color: white;
}
