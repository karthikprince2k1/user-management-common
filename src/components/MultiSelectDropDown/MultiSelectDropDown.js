import React, { useState, useEffect, useRef } from "react";
import "./MultiSelectDropDown.css";

function MultiSelectDropDown(props) {
  const [clicked, setClicked] = useState(false);
  const [selectedItems, setSelectedItems] = useState([]);
  const wrapperRef = useRef(null);
  const handleClick = function (e) {
    setClicked(!clicked);
  };
  const handleOutsideClick = function (event) {
    if (wrapperRef && !wrapperRef.current.contains(event.target)) {
      setClicked(false);
    }
  };

  const handleListItemClick = function (option) {
    if (!selectedItems.includes(option))
      setSelectedItems([...selectedItems, option]);
  };

  const handleRemoveItem = function (option) {
    setSelectedItems(selectedItems.filter((item) => item !== option));
  };
  useEffect(() => {
    document.addEventListener("click", handleOutsideClick);
    return () => {
      document.removeEventListener("click", handleOutsideClick);
    };
  }, []);

  return (
    <div className="dropdown" ref={wrapperRef}>
      <div className="dropdown-menu" onClick={handleClick}>
        {selectedItems.map((item) => {
          return (
            <span className="dropdown-seleced-items">
              {item}{" "}
              <span
                className="close"
                onClick={() => handleRemoveItem(item)}
              ></span>
            </span>
          );
        })}
        <svg
          viewBox="0 0 140 140"
          width="24"
          height="24"
          xmlns="http://www.w3.org/2000/svg"
          className={clicked ? "rotate" : ""}
        >
          <g>
            <path
              d="m121.3,34.6c-1.6-1.6-4.2-1.6-5.8,0l-51,51.1-51.1-51.1c-1.6-1.6-4.2-1.6-5.8,0-1.6,1.6-1.6,4.2 0,5.8l53.9,53.9c0.8,0.8 1.8,1.2 2.9,1.2 1,0 2.1-0.4 2.9-1.2l53.9-53.9c1.7-1.6 1.7-4.2 0.1-5.8z"
              fill="black"
            />
          </g>
        </svg>
      </div>

      <ul
        className={clicked ? "dropdown-list-visible" : "dropdown-list-hidden"}
      >
        {props.options.map((option) => {
          return (
            <li value={option} onClick={() => handleListItemClick(option)}>
              {option}
            </li>
          );
        })}
      </ul>
    </div>
  );
}

export default MultiSelectDropDown;
