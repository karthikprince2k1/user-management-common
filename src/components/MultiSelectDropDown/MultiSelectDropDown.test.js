import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import MultiSelectDropDown from "./MultiSelectDropDown";

afterEach(cleanup);

describe("Multi Select Drop Down Test Suite", () => {
  it("should take a snapshot", () => {
    const { asFragment } = render(<MultiSelectDropDown />);

    expect(asFragment(<MultiSelectDropDown />)).toMatchSnapshot();
  });

  it("should take a snapshot with options", () => {
    const options = ["option1", "option2", "option 3"];
    const { asFragment } = render(<MultiSelectDropDown options={options} />);

    expect(
      asFragment(<MultiSelectDropDown options={options} />)
    ).toMatchSnapshot();
  });

  it("should open the dropdown on clicking the dropdown menu and list the items", () => {
    const options = ["option1", "option2", "option 3"];
    const { getByTestId, getAllByTestId } = render(
      <MultiSelectDropDown options={options} />
    );

    fireEvent.click(getByTestId("dropdown-menu"));

    expect(getByTestId("arrow-down").classList.contains("rotate")).toBe(true);
    expect(
      getByTestId("dropdown-list").classList.contains("dropdown-list-visible")
    ).toBe(true);
    expect(getByTestId("dropdown-list").children.length === 3).toBe(true);

    const listItems = getAllByTestId("dropdown-list-item").map(
      (li) => li.textContent
    );
    listItems.forEach((li, i) => {
      expect(li).toBe(options[i]);
    });
  });

  it("should display the unique selected items from the list even though the same item gets clicked again", () => {
    const options = ["option1", "option2", "option 3", "option 4"];
    const { getByTestId, getAllByTestId } = render(
      <MultiSelectDropDown options={options} />
    );
    fireEvent.click(getByTestId("dropdown-menu"));
    const listItems = getAllByTestId("dropdown-list-item");

    fireEvent.click(listItems[0]);
    fireEvent.click(listItems[3]);
    fireEvent.click(listItems[3]);

    console.log(getByTestId("dropdown-menu").textContent);
    expect(getByTestId("dropdown-menu")).toHaveTextContent(
      options[0] + " " + options[3]
    );
  });

  it("should remove the item from the list on click of x button", () => {
    const options = ["option1", "option2", "option 3", "option 4"];
    const { getByTestId, getAllByTestId } = render(
      <MultiSelectDropDown options={options} />
    );
    fireEvent.click(getByTestId("dropdown-menu"));
    const listItems = getAllByTestId("dropdown-list-item");
    fireEvent.click(listItems[0]);
    fireEvent.click(listItems[3]);

    const item1 = getAllByTestId("remove-item")[0];
    fireEvent.click(item1);

    expect(getByTestId("dropdown-menu")).toHaveTextContent(options[3]);
  });
});
