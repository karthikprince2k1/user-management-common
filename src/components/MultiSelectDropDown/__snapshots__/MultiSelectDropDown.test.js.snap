// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multi Select Drop Down Test Suite should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="dropdown"
  >
    <div
      class="dropdown-menu"
      data-testid="dropdown-menu"
    >
      <svg
        class=""
        data-testid="arrow-down"
        height="24"
        viewBox="0 0 140 140"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="m121.3,34.6c-1.6-1.6-4.2-1.6-5.8,0l-51,51.1-51.1-51.1c-1.6-1.6-4.2-1.6-5.8,0-1.6,1.6-1.6,4.2 0,5.8l53.9,53.9c0.8,0.8 1.8,1.2 2.9,1.2 1,0 2.1-0.4 2.9-1.2l53.9-53.9c1.7-1.6 1.7-4.2 0.1-5.8z"
            fill="black"
          />
        </g>
      </svg>
    </div>
    <ul
      class="dropdown-list-hidden"
      data-testid="dropdown-list"
    />
  </div>
</DocumentFragment>
`;

exports[`Multi Select Drop Down Test Suite should take a snapshot with options 1`] = `
<DocumentFragment>
  <div
    class="dropdown"
  >
    <div
      class="dropdown-menu"
      data-testid="dropdown-menu"
    >
      <svg
        class=""
        data-testid="arrow-down"
        height="24"
        viewBox="0 0 140 140"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="m121.3,34.6c-1.6-1.6-4.2-1.6-5.8,0l-51,51.1-51.1-51.1c-1.6-1.6-4.2-1.6-5.8,0-1.6,1.6-1.6,4.2 0,5.8l53.9,53.9c0.8,0.8 1.8,1.2 2.9,1.2 1,0 2.1-0.4 2.9-1.2l53.9-53.9c1.7-1.6 1.7-4.2 0.1-5.8z"
            fill="black"
          />
        </g>
      </svg>
    </div>
    <ul
      class="dropdown-list-hidden"
      data-testid="dropdown-list"
    >
      <li
        data-testid="dropdown-list-item"
        value="option1"
      >
        option1
      </li>
      <li
        data-testid="dropdown-list-item"
        value="option2"
      >
        option2
      </li>
      <li
        data-testid="dropdown-list-item"
        value="option 3"
      >
        option 3
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
